lib/Google/RestApi.pm

```diff
@@ -32,32 +34,51 @@ sub new {
   my $class = shift;
 
   state $check = compile_named(
-    config_file => Str, { optional => 1 },
-    _extra_     => slurpy Any,
+    config_file          => Str, { optional => 1 },
+    service_account_file => Str, { optional => 1 },
+    _extra_              => slurpy Any,
   );
   my $self = named_extra($check->(@_));
 
+
   if ($self->{config_file}) {
     my $config = eval { LoadFile($self->{config_file}) };
     die "Unable to load config file '$self->{config_file}': $@" if $@;
     $self = Hash::Merge::merge($self, $config);
+  } elsif ($self->{service_account_file}) {
+    die "Unable to find Service Account JSON file at: $self->{service_account_file}" unless -e $self->{service_account_file};
   }
 
+
+
   state $check2 = compile_named(
-    config_file   => Str, { optional => 1 },
-    client_id     => Str,
-    client_secret => Str,
-    token_file    => Str,
-    timeout       => Int, { default => 120 },
-    throttle      => Int->where('$_ > -1'), { default => 0 },
-    post_process  => CodeRef, { optional => 1 },
+    config_file          => Str, { optional => 1 },
+    client_id            => Str, { optional => 1 },
+    client_secret        => Str, { optional => 1 },
+    token_file           => Str, { optional => 1 },
+    service_account_file => Str, { optional => 1 },
+    timeout              => Int, { default => 120 },
+    throttle             => Int->where('$_ > -1'), { default => 0 },
+    post_process         => CodeRef, { optional => 1 },
   );
   $self = $check2->(%$self);
 
-  $self->{token_file} = dirname($self->{config_file}) . "/$self->{token_file}"
-    if !-f $self->{token_file} && $self->{config_file};
-  die "Token file not found: '$self->{token_file}'"
-    if !-f $self->{token_file};
+  if ($self->{service_account_file}) {
+    my $auth = WWW::Google::Cloud::Auth::ServiceAccount->new(
```

Comment:
This should be moved to sub access_token where scope can be passed as an argument. You have hard-coded drive and spreadsheets scope in the ctor here, but this is intended to be generic and support any potential api from google.

Reply:
there is a tiny problem to be resolved:
the scope is set via `bin/google_restapi_session_creator` and stored in the `token_file`.
the `service_account_file` however is generic and does not contain a scope. (it is downloaded directly from google and does not require any changes).

proposed solution:
require the scope to be in the arguments of `RestApi->new()` along with `service_account_file `if the service account auth is used. I have to check if `Type::Params` allows for such complex validation.

what do you think?

Reply:
I think what I will do is accept this pr as is, but with the intention to review the entire auth strategy. If, down the road, a third option arrives (OAuth3?), and then a forth and a fifth, we'll have to jam them all into the existing design, and the code is will become progressively incomprehensible. I think that's because we are making the RestApi class also be responsible for authorization. RestApi should be used to communicate with Google, with a secondary authorization class handling the auth more independently from this module.

So I'd expect another commit later on that helps to extract the auth config from this RestApi module so that the module will be a step removed from whatever auth is being used, and each auth class will plug itself in and modify the rest traffic accordingly.

This will likely be a breaking change, but this is still in beta status...

Let me know whether or not that makes sense.

Reply:
It makes perfect sense.
the 'glue' between the api handling and the auth token provider should not be in either of the two.
the glue is perfectly fine in the program that uses both to fulfill its purpose. things should be well documented though.
i.e. examples for both auth methods.

how about:
requiring a coderef called 'access_token' in the config that provides the access token? after all this token is all the api code needs. 

this would also reduce the dependencies of the rest-api code.

Reply:
I've taken your suggestions here and redesigned the auth mechanism code. Please see the other pull request submitted by me. In particular, check out the perldoc for RestApi regarding the config file for initializing the different auth mechanisms.

The 'glue' between RestApi and the various auth mechs is that RestApi will call method 'headers' on the given auth mech and add any returned headers to the Furl init call. Otherwise, RestApi couldn't care less how the headers are generated.

This greatly simplifies RestApi and allows flexible creation of authorization objects, and should meet our future needs.

Please let me know if you have any questions/concerns with how I've implemented this. Feel free to modify the service account code, add tests, update the perldoc, add you as an author etc, either here or after this has been merged.

Thanks for bringing this up.

Reply:
I like the design. however, it still keeps auth handling in the Google::RestApi package. I was under the impression that it should be separated. be that as it may, it is fine for me.

config is simple and nice:
```
my $rest = Google::RestApi->new(
    auth => {
        class=> 'ServiceAccount',
        account_file=>'/path/to/service_account.json',
        scope => [qw{
            https://www.googleapis.com/auth/drive
            https://www.googleapis.com/auth/spreadsheets
        }],
    },
);
```

I did some changes to let WWW::Google::Cloud::Auth::ServiceAccount handle the token caching and expiry. with your permission I will push them to your branch.

Reply:
I'm not sure how you would completely remove the auth code from the package and still allow RestApi to know how to deal with each one. The Auth* classes are intended to wrap around whatever auth mech you choose and create a standard call that RestApi can always rely on (sub 'headers'). 

Also, there is nothing to stop you from creating your own auth outside the package all together, you would just have to instantiate it, and pass it as 'auth => $object' (instead of the hash), as long as you provide a 'headers' routine.

Absolutely push whatever changes you like to the other branch. I just made my best guess on how to use ServiceAccount, but didn't expect it would be all correct.

Reply:
like this:

```
use WWW::Google::Cloud::Auth::ServiceAccount;

my $auth = WWW::Google::Cloud::Auth::ServiceAccount->new(
    credentials_path => '/home/myapp/priv/google_some_service.json',
);

my $rest = Google::RestApi->new(
    auth_teaders => sub{
        return { 
            'Authorization' => 'Bearer ' . $auth->get_token,    },
        },  
);   
``` 
Google::RestApi only calls the sub every time it needs headers.
caching is done in the sub or in this case WWW::Google::Cloud::Auth::ServiceAccount handles it.

both approaches have merit. and they aren't even mutually exclusive. 
Google::RestApi::Auth::* is more structured and elegant
the coderef way allows for less reading and understanding before writing the code.
but let's not over-engineer this and stick to the Google::RestApi::Auth::* way.
I have already modified it to let WWW::Google::Cloud::Auth::ServiceAccount do the caching.

> Absolutely push whatever changes you like to the other branch. I just made my best guess on how to use ServiceAccount, but didn't expect it would be all correct.

I seem to lack the permissions to do so. Either this or my git setup is broken.

Reply:
Ah right, I see what you meant now in your earlier post. I hadn't quite understood.

One of the things I wanted to avoid was having to do a lot of initialization code each time you use the framework in your scripts, keep it to just initializing RestApi in one statement and carry on. As it is, you can just 'my $api = RestApi->new(config_file => $path)', and you are good to go. I initially thought of passing some external auth object in, but figured that was adding too much complexity for each script you wanted to use it with, forcing each user to create some common init module, or worse, cutting and pasting several lines of code for each script.

Let me see if I can change any permissions on my end.

Reply:
I added you as a collaborator, you should get an invite. Once you accept the invite, I think I can assign you to the other branch.